subsys/bluetooth/mesh/shell/CMakeLists.txt

```diff
@@ -5,6 +5,9 @@ zephyr_library_sources(
 	utils.c
 	)
 
+# Adds bt_shell_private.c if CONFIG_BT_SHELL is not defined, as it is already built when enabled.
+zephyr_library_sources_ifndef(CONFIG_BT_SHELL ../../host/shell/bt_shell_private.c)
+
```

Comment:
If this API is being used not just by host, then I think it should be moved to `common/` folder.

This `zephyr_library_sources_ifndef` part can be fixed by a hidden Kconfig option (added to `common/Kconfig`) which is selected by `BT_MESH` or `BT_SHELL` Kconfig options. If I'm not missing anything, this will also solve the `_ifndef` part.

Reply:
Thank you for the suggestion. Here's the updated approach:

common/Kconfig
```
config BT_PRIVATE_SHELL
    bool
    default y if BT_MESH_SHELL || BT_SHELL
```

common/CMakeList
```
zephyr_library_sources_ifdef(CONFIG_BT_PRIVATE_SHELL bt_shell_private.c)
```

Then move `bt_shell_private.c` and `bt_shell_private.h` into the `common` folder instead.

WDYT?

Reply:
Looks good.  But I'd add `select BT_PRIVATE_SHELL` to `BT_MESH_SHELL` and `BT_SHELL` so it is unconditionally selected when one of these options is enabled. This also makes visible dependency of host and mesh shell on that API.

Reply:
What is the use case where `BT_MESH_SHELL=y` and `BT_SHELL=n`? 

Shouldn't `BT_MESH_SHELL` depend on `BT_SHELL`, so that `bt_shell_private` is always compiled for `BT_MESH_SHELL`? 

Do we need `BT_MESH_SHELL` at all, or should it just be similar to LE, LL and Classic and enable the mesh shell if `BT_SHELL && BT_MESH`?

Reply:
`BT_MESH_SHELL` is completely independent of `BT_SHELL`:
https://github.com/zephyrproject-rtos/zephyr/blob/ef6f804d8f2648935afadaab09b511913ffd9f4c/subsys/bluetooth/mesh/shell/Kconfig#L6-L11

And currently they can't work together. Mesh must be initialized before settings are loaded. The bt init function doesn't do this.

Reply:
> And currently they can't work together. Mesh must be initialized before settings are loaded. The the shell's bt init function doesn't do this.

That sounds fairly trivial to implement. 

Would it make sense to get rid of `BT_MESH_SHELL` (not in this PR of course)?

Reply:
No, it is not trivial as it depends on composition data and provisioning context.

There is no such request from users/customers/PMT. Also, 2 shells seem to have different purpose. For mesh most of the code is embedded into an application and the application can enable the mesh shell only for a certain model. For example, you can enable shell commands for client models without changing code in your application by toggling few Kconfig options. The host's shell adds a lot of code which is not needed in most (if not all) mesh applications.

Reply:
Hi @Thalley, according to @PavelVPV's response, can I move forward with the changes as per his suggestion?
Also, I need to resolve the merge conflict too. 😃